File: Backend/src/main/java/lan/dk/podcastserver/manager/worker/updater/MyCanalUpdater.java
package lan.dk.podcastserver.manager.worker.updater;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.jayway.jsonpath.TypeRef;
import io.vavr.collection.HashSet;
import io.vavr.collection.List;
import io.vavr.collection.Set;
import io.vavr.control.Option;
import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.manager.worker.finder.MyCanalFinder;
import lan.dk.podcastserver.service.HtmlService;
import lan.dk.podcastserver.service.ImageService;
import lan.dk.podcastserver.service.JsonService;
import lan.dk.podcastserver.service.SignatureService;
import lan.dk.podcastserver.service.properties.PodcastServerParameters;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import javax.validation.Validator;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

import static io.vavr.API.Option;
import static io.vavr.API.Tuple;
import static org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_PROTOTYPE;

@Slf4j
@Component
@Scope(SCOPE_PROTOTYPE)
public class MyCanalUpdater extends AbstractUpdater {

    private static final String MYCANAL_DATE_PATTERN = "dd/MM/yyyy-HH:mm:ss";
    private static final TypeRef<Set<MyCanalItem>> SET_OF_MY_CANAL_ITEM = new TypeRef<Set<MyCanalItem>>(){};
    private static final String URL_DETAILS = "https://secure-service.canal-plus.com/video/rest/getVideosLiees/cplus/%s?format=json";
    private static final String DOMAIN = "https://www.mycanal.fr";

    private final JsonService jsonService;
    private final ImageService imageService;
    private final HtmlService htmlService;

    protected MyCanalUpdater(PodcastServerParameters podcastServerParameters, SignatureService signatureService, Validator validator, JsonService jsonService, ImageService imageService, HtmlService htmlService) {
        super(podcastServerParameters, signatureService, validator);
        this.jsonService = jsonService;
        this.imageService = imageService;
        this.htmlService = htmlService;
    }

    @Override
    public Set<Item> getItems(Podcast p) {
        return itemsAsJsonFrom(p)
                .getOrElse(HashSet::empty).toMap(v -> Tuple(v, this.findDetails(v)))
                .filterValues(Option::isDefined)
                .mapValues(Option::get)
                .mapValues(this::asItem)
                .toSet()
                .map(v -> v._2().url(DOMAIN + v._1().getOnClick().getPath()).build());
    }

    @Override
    public String signatureOf(Podcast p) {
        return itemsAsJsonFrom(p)
                .map(items -> items.toList().map(MyCanalItem::getContentID).sorted().mkString())
                .map(signatureService::generateMD5Signature)
                .getOrElseThrow(() -> new RuntimeException("Error during signature of " + p.getTitle() + " with url " + p.getUrl()));
    }

    private Option<Set<MyCanalItem>> itemsAsJsonFrom(Podcast p) {
        return htmlService
                .get(p.getUrl())
                .map(Document::body)
                .map(es -> List.ofAll(es.select("script")))
                .getOrElse(List::empty)
                .find(e -> e.html().contains("__data"))
                .map(Element::html)
                .flatMap(MyCanalFinder::extractJsonConfig)
                .map(jsonService::parse)
                .map(JsonService.to("landing.strates[0].contents", SET_OF_MY_CANAL_ITEM))
                ;
    }

    private Option<MyCanalDetailsItem> findDetails(MyCanalItem item) {
        return jsonService.parseUrl(String.format(URL_DETAILS, item.getContentID()))
                .map(JsonService.to(MyCanalDetailsItem.class))
                ;
    }

    @Override
    public Type type() {
        return new Type("MyCanal", "MyCanal");
    }

    @Override
    public Integer compatibility(String url) {
        return MyCanalFinder._compatibility(url);
    }

    private Item.ItemBuilder asItem(MyCanalDetailsItem i) {
        MyCanalInfosItem infos = i.getInfos();
        MyCanalTitrageItem titrage = infos.getTitrage();
        MyCanalPublicationItem publication = infos.getPublication();

        return Item.builder()
                .title(titrage.getTitre())
                .description(titrage.getSous_titre())
                .length(i.getDuration())
                .cover(i.getMedia().getImages().cover().map(imageService::getCoverFromURL).getOrElse(Cover.DEFAULT_COVER))
                .pubDate(publication.asZonedDateTime())
                ;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class MyCanalItem {
        @Getter @Setter String contentID;
        @Getter @Setter MyCanalItemOnClick onClick;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class MyCanalItemOnClick {
        @Getter @Setter String path;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class MyCanalDetailsItem {
        @JsonProperty("ID") @Getter @Setter String id;
        @JsonProperty("DURATION") @Getter @Setter Long duration;
        @JsonProperty("INFOS") @Getter @Setter MyCanalInfosItem infos;
        @JsonProperty("MEDIA") @Getter @Setter MyCanalMediaItem media;
        @JsonProperty("URL") @Getter @Setter String url;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class MyCanalInfosItem {
        @JsonProperty("DESCRIPTION") @Getter @Setter String description;
        @JsonProperty("PUBLICATION") @Getter @Setter MyCanalPublicationItem publication;
        @JsonProperty("TITRAGE") @Getter @Setter MyCanalTitrageItem titrage;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class MyCanalPublicationItem {
        @JsonProperty("DATE") @Getter @Setter String date;
        @JsonProperty("HEURE") @Getter @Setter String heure;

        ZonedDateTime asZonedDateTime() {
            LocalDateTime localDateTime = LocalDateTime.parse(date.concat("-").concat(heure), DateTimeFormatter.ofPattern(MYCANAL_DATE_PATTERN));
            return ZonedDateTime.of(localDateTime, ZoneId.of("Europe/Paris"));
        }
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class MyCanalTitrageItem {
        @JsonProperty("TITRE") @Getter @Setter String titre;
        @JsonProperty("SOUS_TITRE") @Getter @Setter String sous_titre;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class MyCanalMediaItem {
        @JsonProperty("IMAGES") @Getter @Setter MyCanalImageItem images;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class MyCanalImageItem {
        @JsonProperty("GRAND") @Getter @Setter String grand;
        @JsonProperty("PETIT") @Getter @Setter String petit;

        Option<String> cover() {
            return Option(grand).orElse(Option(petit));
        }
    }


}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/worker/mycanal/MyCanalUpdater.kt
package com.github.davinkevin.podcastserver.manager.worker.mycanal

import arrow.core.*
import arrow.syntax.collections.firstOption
import com.github.davinkevin.podcastserver.manager.worker.*
import com.github.davinkevin.podcastserver.service.HtmlService
import com.github.davinkevin.podcastserver.service.ImageService
import com.github.davinkevin.podcastserver.service.SignatureService
import com.github.davinkevin.podcastserver.utils.toTry
import com.jayway.jsonpath.TypeRef
import lan.dk.podcastserver.service.JsonService
import lan.dk.podcastserver.service.JsonService.to
import org.jsoup.select.Elements
import org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_PROTOTYPE
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component
import java.net.URI

@Component
@Scope(SCOPE_PROTOTYPE)
class MyCanalUpdater(val signatureService: SignatureService, val jsonService: JsonService, val imageService: ImageService, val htmlService: HtmlService) : Updater {

    override fun blockingFindItems(podcast: PodcastToUpdate): Set<ItemFromUpdate> =
            itemsAsJsonFrom(URI(podcast.url.toASCIIString()))
                    .getOrElse { setOf() }
                    .flatMap { findDetails(it).toList() }
                    .map { toItem(it.second, DOMAIN + it.first.onClick.path) }
                    .toSet()

    override fun blockingSignatureOf(url: URI): String {
        return itemsAsJsonFrom(url)
                .map { items -> items.map { it.contentID }.sorted().joinToString(",") }
                .map { signatureService.fromText(it) }
                .getOrElse { throw RuntimeException("Error during signature of podcast with url " + url.toASCIIString()) }
    }

    private fun itemsAsJsonFrom(url: URI) =
            htmlService.get(url.toASCIIString())
                    .map { it.body() }
                    .map { it.select("script") }
                    .getOrElse { Elements() }
                    .asSequence()
                    .map { it.html() }
                    .firstOption { "__data" in it }
                    .flatMap { extractJsonConfig(it).toOption() }
                    .map { jsonService.parse(it) }
                    .map { to("landing.strates[*].contents[*]", SET_OF_MY_CANAL_ITEM).apply(it) }

    private fun findDetails(item: MyCanalItem): Option<Pair<MyCanalItem, MyCanalDetailsItem>> {
        val json = jsonService.parseUrl(URL_DETAILS.format(item.contentID)).toTry()

        val fromCollection = {
            json
                    .flatMap { Try { JsonService.to(SET_OF_MY_CANAL_DETAILS_ITEM).apply(it) } }
                    .getOrElse { setOf() }
                    .firstOption { it.id == item.contentID }
        }

        return json
                .flatMap { Try { JsonService.to(MyCanalDetailsItem::class.java).apply(it) } }
                .toOption()
                .orElse(fromCollection)
                .map { item to it }

    }

    override fun type() = Type("MyCanal", "MyCanal")

    override fun compatibility(url: String?) = myCanalCompatibility(url)

    private fun toItem(i: MyCanalDetailsItem, url: String): ItemFromUpdate {
        val infos = i.infos
        val titrage = infos.titrage
        val publication = infos.publication

        return ItemFromUpdate(
            title = titrage.titre,
            description = titrage.sous_titre,
            length = i.duration,
            cover = i.media.images.cover().map { imageService.fetchCoverInformation(it) }.getOrElse { null }?.toCoverFromUpdate(),
            pubDate = publication.asZonedDateTime(),
            url = URI(url)
        )
    }

    companion object {

        private val SET_OF_MY_CANAL_ITEM = object : TypeRef<Set<MyCanalItem>>() {}
        private val SET_OF_MY_CANAL_DETAILS_ITEM = object : TypeRef<Set<MyCanalDetailsItem>>() {}

        private const val URL_DETAILS = "https://secure-service.canal-plus.com/video/rest/getVideosLiees/cplus/%s?format=json"
        private const val DOMAIN = "https://www.mycanal.fr"
    }
}
